compiler/parser/src/common.rs

```diff
@@ -0,0 +1,939 @@
+                    item.span = span;
+
+                    return Ok(item);
+                }
+            }
+
+            elements.push(item);
+
+            if bracket.is_some() {
+                break;
+            }
+        }
+
+        Ok(ast::Item::new(
+            default(),
+            span.merge_into(&path),
+            ast::SequenceLiteral {
+                path,
+                elements: Spanned::new(span, elements),
+            }
+            .into(),
+        ))
+    }
+
+    /// Finish parsing a [record literal] given the already parsed leading path and the
+    /// span of the already parsed leading `{`.
+    ///
+    /// # Grammar
+    ///
+    /// ```grammar
+    /// Record-Literal-⟪Item⟫ ::=
+    ///     "{"
+    ///     ; @Task generalize #Word to "paths" (i.e. identifiers with "." OR "::" seps!)
+    ///     (#Word ("=" ⟪Item⟫)? ",")* ; @Task make trailing comma optional & don't allow it if ";" follows
+    ///     (";" ⟪Item⟫)?
+    ///     "}"
+    /// ```
+    ///
+    /// [record literal]: ast::RecordLiteral
+    pub(crate) fn finish_parse_record_literal<T>(
+        &mut self,
+        path: Option<ast::Path>,
+        mut span: Span,
+    ) -> Result<ast::Item<T>>
+    where
+        T: Item + From<ast::RecordLiteral<ast::Item<T>>>,
+    {
+        let mut fields = Vec::new();
+        let mut base = None;
+
+        loop {
+            if let Some(bracket) = self.consume_span(ClosingCurlyBracket) {
+                span.merging(bracket);
+                break;
+            }
+
+            let name = self.parse_word()?;
+            let mut item = None;
+
+            if self.consume(Equals) {
+                item = Some(self.parse_item()?);
+            }
+
+            fields.push(ast::Field { name, item });
+
+            if self.consume(Semicolon) {
+                base = Some(self.parse_item()?);
+                span.merging(self.expect(ClosingCurlyBracket)?);
+                break;
+            }
+
+            if let Some(bracket) = self.consume_span(ClosingCurlyBracket) {
+                span.merging(bracket);
+                break;
+            }
+
+            self.expect(Comma)?;
+        }
+
+        Ok(ast::Item::new(
+            default(),
+            span,
+            ast::RecordLiteral {
+                path,
+                fields: Spanned::new(span, fields),
+                base,
+            }
+            .into(),
+        ))
+    }
+
+    /// Parse an [application] or a lower item.
+    ///
+    /// # Grammar
+    ///
+    /// ```grammar
+    /// Application-⟪Item⟫-Or-Lower ::= Lower-⟪Item⟫ ⟪Item⟫-Argument*
+    /// ⟪Item⟫-Argument ::=
+    ///     "'"?
+    ///     (Lower-⟪Item⟫ | "(" (#Word "=")? ⟪Item⟫ ")")
+    /// ```
+    ///
+    /// [application]: ast::Application
+    // @Task update grammar
+    pub(crate) fn parse_application_or_lower<T>(&mut self) -> Result<ast::Item<T>>
+    where
+        T: Item + From<ast::Application<ast::Item<T>>> + 'static,
```

Comment:
```suggestion
        T: Item + From<ast::Application<ast::Item<T>>>
```

This can go again, right? We're no longer using a closure for error annotations (we've defunctionalized them)